src/wiki/plugins/images/markdown_extensions.py

```diff
@@ -1,106 +1,84 @@
-                    width = size.split("x")[0] if size else None
-                    html = render_to_string(
-                        "wiki/plugins/images/render.html",
-                        context={
-                            'image': image,
-                            'caption': caption_placeholder,
-                            'align': alignment,
-                            'size': size,
-                            'width': width
-                        })
-                    html_before, html_after = html.split(caption_placeholder)
-                    placeholder_before = self.markdown.htmlStash.store(
-                        html_before,
-                        safe=True)
-                    placeholder_after = self.markdown.htmlStash.store(
-                        html_after,
-                        safe=True)
-                    new_line = placeholder_before + "\n".join(
-                        caption_lines) + placeholder_after + "\n"
-                    previous_line_was_image = False
-                    if previous_line is not "":
-                        if previous_line[line_index:] is not "":
-                            new_line = new_line[0:-1]
-                        new_text[-1] = (previous_line[0:line_index] +
-                                        new_line +
-                                        previous_line[line_index:] +
-                                        "\n" +
-                                        line)
-                        line = None
-                    else:
-                        line = new_line + line
-            if line is not None:
-                new_text.append(line)
-        return new_text
+        size = settings.THUMBNAIL_SIZES["default"]
+
+        image_id = m.group("id").strip()
+        alignment = m.group("align")
+        if m.group("size"):
+            size = settings.THUMBNAIL_SIZES[m.group("size")]
+        try:
+            image = models.Image.objects.get(
+                article=self.markdown.article,
+                id=image_id,
+                current_revision__deleted=False,
+            )
+        except models.Image.DoesNotExist:
+            pass
+
+        caption = m.group("caption")
+
+        caption_placeholder = "{{{IMAGECAPTION}}}"
+        width = size.split("x")[0] if size else None
+        html = render_to_string(
+            "wiki/plugins/images/render.html",
+            context={
+                "image": image,
+                "caption": caption_placeholder,
+                "align": alignment,
+                "size": size,
+                "width": width,
+            },
+        )
+        html_before, html_after = html.split(caption_placeholder)
+        placeholder_before = self.markdown.htmlStash.store(html_before, safe=True)
+        placeholder_after = self.markdown.htmlStash.store(html_after, safe=True)
+        return placeholder_before + caption + placeholder_after
```

Comment:
This is such a great example of how doing things "the right way" is rewarding: Readability, simplicity, performance etc..